### ISSUE TITLE
Polymorphic serialization: use property name directly as type info, without wrapper

### ISSUE THREAD

**Issue by fransflippo** [FasterXML/jackson-databind#2005]
Given:
    public class Zoo {
      private Animal animal;
    }
    public abstract class Animal {
      private String name = "Nenji";
    }
    public class Dog extends Animal {
    }
    public class Cat extends Animal {
    }

I want to be able to serialize an instance of `Zoo` so that it outputs either:
    {
      "dog": {
        "name": "name"
      }
    }

or

    {
      "cat": {
        "name": "name"
      }
    }

depending on the concrete type of `animal` in `Zoo`.

As far as I can tell, none of the options `@JsonTypeInfo` currently support this. The closest I can get is with the default `@JsonTypeInfo.As.WRAPPER_OBJECT`, but that results in:

    {
      "animal":
        "cat": {
          "name": "name"
        }
      }
    }

I really don't want to have the wrapper object. The reason is that I want to remain backwards compatible with a previous incarnation of my object model where there were only `Dog`s, the `Animal` abstraction hadn't been created, and `Animal animal` in `Zoo` was `Dog dog`.

Is it possible to add a sub-type (de)serialization option that implements this?

**Comment by cowtowncoder** [comment]
No, none of the options work that way and `@JsonTypeInfo` is not really designed to support dynamically varying property names like that (or, `BeanDeserializer` in general).

So for this you'd have to implement your own handling via custom `JsonDeserializer`.

**Comment by fransflippo** [comment]
It would make sense to have this work since, if using the `@XmlElements` annotation to indicate the sub-types instead of the Jackson annotations, it *will* serialize that way to XML (i.e. outputting either a `<dog>` or a `<cat>`, but not to JSON (since the wrapper will be output).

I'll look at implementing this with a custom serializer for now.

Thanks!
Frans

**Comment by cowtowncoder** [comment]
@fransflippo valid point, wrt XML, since wrapping on that side is... bit different. Might even be more doable for that module actually, since various work-arounds are already needed (input stream is heavily massaged prior to being exposed to databinding).
